### PR TITLE
Sharing

### LINKED ISSUE #7: Implement Sharing
- [ ] complete implementation for `davCollectionShareable`
- [ ] implement `principalPropertySearch` to search for users and groups

### LINKED ISSUE #8: Implement Publishing
- [ ] complete implementation for `davCollectionPublishable`

### PR DESCRIPTION
**Pull request by georgehrke**
fixes #7 
fixes #8

### CODE REVIEW

**Comment by skjnldsv** [comment]
What's left to do here? :)

**Comment by georgehrke** [comment]
Just a couple of remarks how to use this:
- Do not use this OCS Share API like [it's currently done in Contacts](https://github.com/nextcloud/contacts/blob/38ea18f362d5c506582cfa715f30789fb85beae3/src/components/Settings/SettingsAddressbookShare.vue#L151L154), but use `client.principalPropertySearchByDisplayname(...)` instead
- it will return an array of `Principal`s.
- you have to filter the principals by their `calendarUserType`. You only want a `calendarUserType` of `INDIVIDUAL` or `GROUP` (otherwise rooms and resources will also show up, which is not wanted)
- to share a calendar / address book, you can't just use the userid, but you need a special uri that looks like this: `principal:principals/users/<userid>` respectively `principal:principals/groups/<groupid>`
- every `Principal` object has a property `principalScheme` that contains exactly this special uri.
- `Principal` objects also have the `displayname` property that contains the user's/group's display name.

So the workflow is the following:
- as the user types, use `client.principalPropertySearchByDisplayname(...)` to gather the principals, display them to the user, let them select.
- Use `principal.principalScheme` as the first parameter of `addressbook.share`
- For now it is the app's responsibility to store the sharee's displayname. The `common-name` property of shares will only be populated after a complete refetch of the collection's properties.
- If you want to update an share's read-write access, simply share it as if it was a new share.

**Comment by georgehrke** [comment]
I added a `userId` (etc...) property to `Principal`, so that it's easier to use Principals with the avatar component.

**Comment by skjnldsv** [comment]
On it! :muscle:

File: src/models/principal.js [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+	 * as specified in RFC 3744
+	 *
+	 * https://tools.ietf.org/html/rfc3744#section-2
+	 *
+	 * @inheritDoc
+	 */
+	constructor(...args) {
+		super(...args);
+
+		super._exposeProperty('displayname', NS.DAV, 'displayname');
```
Comment (skjnldsv) [inline src/models/principal.js]: We use pascalCase everywhere but here, should we?
`displayName`

**Comment by skjnldsv** [comment]
@georgehrke how do I change the permission of a share? :)

File: src/models/principal.js (hunk 1 above)
Comment (georgehrke) [reply src/models/principal.js]: It's `displayname` in `DavCollection.js` as well, so i wanted to keep it consistent:
https://github.com/nextcloud/cdav-library/blob/master/src/models/davCollection.js#L70

**Comment by georgehrke** [comment]
> could be nice to have the userid

Should I add some helper methods to the DavClient class that will tell you whether or not a `principal://...` uri is a user and will extract the userid? 

> and displayname with the initial PROPFIND

Displayname is stored in `common-name`. (just sticking to the CalDAV names here)


> what use the 'invite-accepted' key have?

Right now we are using our own hacky sharing standard. `invite-accepted` is returned by the server for every PROPFIND, but it's always true.
Eventually it would be very cool if we could implement the official sharing standard. There are more values described in that standard.
https://github.com/apple/ccs-calendarserver/blob/master/doc/Extensions/caldav-sharing.txt

@schiessle and me are gonna discuss this standard and evaluate it in the context of federated sharing at the next hackweek. You are very welcome to join 😀

**Comment by skjnldsv** [comment]
> Displayname is stored in `common-name`. (just sticking to the CalDAV names here)

Not for groups :)

> Should I add some helper methods to the DavClient class that will tell you whether or not a `principal://...` uri is a user and will extract the userid?

I'm manually doing that right now, but could be a nice to have this built-in :)

**Pull request merged**